src/Parsers/FunctionSecretArgumentsFinder.h

```diff
@@ -29,6 +32,21 @@ class AbstractFunction
         virtual ~Arguments() = default;
         virtual size_t size() const = 0;
         virtual std::unique_ptr<Argument> at(size_t n) const = 0;
+        void skipArgument(size_t n) { skipped_indexes.insert(n); }
+        void unskipArguments() { skipped_indexes.clear(); }
+        size_t getRealIndex(size_t n) const
+        {
+            for (auto idx : skipped_indexes)
+            {
+                if (n < idx)
+                    break;
+                ++n;
+            }
+            return n;
+        }
+        size_t skippedSize() const { return skipped_indexes.size(); }
+    private:
+        std::set<size_t> skipped_indexes;
```

Comment:
This is kind of confusing. I get the idea, but not entirely and why it has to be done this way.

Why do you have to manually call `getRealIndex` in some places knowing you already implemented that in ARgumentsAST:

```
std::unique_ptr<Argument> at(size_t n) const override
        { /// n is relative index, some can be skipped
            return std::make_unique<ArgumentTreeNode>(arguments->at(getRealIndex(n)).get());
        }
```

Why unskip?

Reply:
In `markSecretArgument` and `maskAzureConnectionString` index goes `outside` in code, which works with raw index, not through this local `Arguments` class.
Like here - https://github.com/ClickHouse/ClickHouse/blob/master/src/Analyzer/Resolve/QueryAnalyzer.cpp#L2932